src/common/pipeline/pipelines.hpp

```diff
@@ -312,7 +312,9 @@ class ModifyingPipeline : public Pipeline<T, T, N> {
         }
         *this->product = input;
         for (size_t i = 0; i < this->size; i++) {
-            dynamic_cast<ModifyingStage<T> *>(this->stages[i])->SetResource(*this->product);
+            auto *stage = static_cast<ModifyingStage<T> *>(this->stages[i]);
+            assert(stage != nullptr);
```

Comment:
This check, although admirable, is impractical in nominal use cases, please get rid of the assertion and go straight for the `SetResource` operation (also meaning you don't need to store a variable)

I'm also interested to hear why you chose a `static_cast` instead of `dynamic_cast`. I think it could be justified, just as long as we also agree on it.